### ISSUE TITLE
When peft>0.7.0, fine-tuning ChatGLM3-6B causes the model to become dumb with a loss of 0

### ISSUE THREAD

**Issue by Tangent-90C** [huggingface/peft#1568]
### System Info

```
Package                        Version
------------------------------ ---------------
accelerate                     0.28.0
datasets                       2.18.0
peft                           0.9.0
torch                          2.1.2+cu121
transformers                   4.38.2
```

### Who can help?

@pacman100 @younesbelkada @sayakpaul

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

This is my custom dataset:
https://github.com/datawhalechina/self-llm/blob/master/dataset/huanhuan.json

Here's the code to reproduce it
```
import torch
from datasets import Dataset
from transformers import AutoTokenizer, AutoModelForCausalLM, DataCollatorForSeq2Seq, TrainingArguments, Trainer
import pandas as pd
from peft import TaskType, get_peft_model, LoraConfig

model_name = "THUDM/chatglm3-6b"

# 数据处理流程,参考GLM3仓库:https://github.com/THUDM/ChatGLM3/blob/main/finetune_chatmodel_demo/preprocess_utils
def process_func(example):
    MAX_LENGTH = 512
    input_ids, labels = [], []
    instruction = tokenizer.encode(text="\n".join(["<|system|>", "现在你要扮演皇帝身边的女人--甄嬛", "<|user|>",
                                    example["instruction"] + example["input"] + "<|assistant|>"]).strip() + "\n",
                                    add_special_tokens=True, truncation=True, max_length=MAX_LENGTH)
    response = tokenizer.encode(text=example["output"], add_special_tokens=False, truncation=True, max_length=MAX_LENGTH)
    
    input_ids = instruction + response + [tokenizer.eos_token_id]
    labels = [tokenizer.pad_token_id] * len(instruction) + response + [tokenizer.eos_token_id]

    pad_len = MAX_LENGTH - len(input_ids)
    input_ids += [tokenizer.pad_token_id] * pad_len
    labels += [tokenizer.pad_token_id] * pad_len
    labels = [(l if l != tokenizer.pad_token_id else -100) for l in labels]

    return {
        "input_ids": input_ids,
        "labels": labels,
    }

args = TrainingArguments(
    output_dir="./autodl-tmp/ChatHUA",
    per_device_train_batch_size=1,
    gradient_accumulation_steps=8,
    logging_steps=20,
    num_train_epochs=1
)

# 将JSON文件转换为CSV文件,处理数据集
df = pd.read_json('./huanhuan.json')
ds = Dataset.from_pandas(df)
# 加载tokenizer
tokenizer =  AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
# 将数据集变化为token形式
tokenized_ds = ds.map(process_func, remove_columns=ds.column_names)

# 创建模型
model = AutoModelForCausalLM.from_pretrained(model_name, trust_remote_code=True,
                                         low_cpu_mem_usage=True, device_map="cuda",
                                         )

# 创建loRA参数
config = LoraConfig(task_type=TaskType.CAUSAL_LM, target_modules={"query_key_value"}, r=8, lora_alpha=32)

# 模型合并
model = get_peft_model(model, config)

# 指定GLM的Data collator
data_collator = DataCollatorForSeq2Seq(
    tokenizer,
    model=model,
    label_pad_token_id=-100,
    pad_to_multiple_of=None,
    padding=False
)

# 指定训练参数。
trainer = Trainer(
    model=model,
    args=args,
    train_dataset=tokenized_ds,
    data_collator=data_collator,
)

# 开始训练
trainer.train()


# Test the fine-tuned model
input_text = '写一篇300字的作文，内容主题自拟'
tokenized_input = tokenizer.build_chat_input(input_text, history=[], role='user').to('cuda')

data = model.generate(**tokenized_input, max_length=512)

print(data)

print(tokenizer.decode(data[0]))
```

### Expected behavior

The fine-tuning model should produce an output anyway

![image](https://github.com/huggingface/peft/assets/28804414/72390b6b-542c-4804-b350-614720ebfc9d)

However, after fine-tuning, the model can't say a word, that is, it can't output a token. In addition, when fine-tuning, the loss will be very large for the first time, and then the loss will be reduced to 0 for all the subsequent losses

![image](https://github.com/huggingface/peft/assets/28804414/6143f44e-5ccd-4272-bb7c-c7100c9d216e)

![image](https://github.com/huggingface/peft/assets/28804414/d0354500-401a-43e5-9f01-bdb26a05f51c)

These issues occurred when I used the latest environment, see System Info.

This issue was resolved when I downgraded peft to 0.6.0.

**Comment by BenjaminBossan** [comment]
This is very hard to diagnose, and not speaking mandarin (?) doesn't help. How long does it take for you to obtain the result? Could you maybe retry with all PEFT releases between 0.6.0 and 0.9.0 so that we know the exact version that causes the problem?

**Comment by Tangent-90C** [comment]
After my tests, to be precise in, This problem occurs when peft>=0.7.0.
Not only Chinese, any token will not be output. This BUG is independent of the dataset you're using. You can reproduce it with any dataset you want.

**Comment by Tangent-90C** [comment]
Yes, this dataset is publicly available

**Comment by BenjaminBossan** [comment]
> Yes, this dataset is publicly available

Is it this one? https://github.com/datawhalechina/self-llm/blob/master/dataset/huanhuan.json

**Comment by Tangent-90C** [comment]
yes

**Comment by BenjaminBossan** [comment]
Okay, I tried to run your script but unfortunately ran out of GPU memory (T4). Even with 4bit weights, I couldn't get this to run. Do you have a smaller model that can reproduce the error?

**Comment by Tangent-90C** [comment]
Sorry, ChatGLM3-6B is already the smallest model and I don't have a GPU to fine-tune it, so I rented a GPU from AutoDL to fine-tune it, it only needs 24GB of memory, you can also try renting a GPU to reproduce bugs.

**Comment by BenjaminBossan** [comment]
Sorry for the delay, I've got access to a machine now to test this. I could reproduce the issue, loss is normal with v0.6.2 and breaks with v0.7.0. Things start to go wrong after the first optimizer step (i.e. the first 8 losses look fine, since we have 8 gradient accumulation steps). Exploding gradients are not the issue: I checked the PEFT parameters and their gradients and they all look good, and choosing a very small learning rate does not help.

This is all I could figure out so far, the root cause is still a mystery to me. I'll investigate more later.

**Comment by BenjaminBossan** [comment]
More updates, I could narrow down the PR that causes this to #1106. Unfortunately, it's a very big refactor, so I don't know what specific change therein is the root cause.

However, I did try running your script using `bfloat16` instead of `float16` and it worked. For this, pass `torch_dtype=torch.bfloat16` to `AutoModelForCausalLM.from_pretrained` and also set `bf16=True` for the `TrainingArguments`. If you could test this to confirm if it works, it would be great.

**Comment by BenjaminBossan** [comment]
Some updates: The code you posted will result in float16 (fp16, half precision) model weights, which causes the issue. You can move to bfloat16 if possible and the error should vanish.

If you can't use bfloat16 (e.g. old GPU), add this:

```python
...
model = get_peft_model(model, config)
# convert all peft parameters to float32
for param in model.parameters():
    if param.requires_grad:
        param.data = param.data.float()
```

**Comment by Tangent-90C** [comment]
Thanks, I just changed the model weights to be bfloat16 type and the problem is solved

**Comment by BenjaminBossan** [comment]
Fantastic, good luck with your further training.
